cmd/generate/main.go

```diff
@@ -0,0 +1,136 @@
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+
+	"github.com/os-observability/tempo-operator/apis/tempo/v1alpha1"
+	"github.com/os-observability/tempo-operator/internal/manifests"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configtempov1alpha1 "github.com/os-observability/tempo-operator/apis/config/v1alpha1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+}
+
+func loadSpec(path string) (v1alpha1.Microservices, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return v1alpha1.Microservices{}, err
+	}
+	defer file.Close()
+
+	spec := v1alpha1.Microservices{}
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 8192)
+	err = decoder.Decode(&spec)
+	if err != nil {
+		return v1alpha1.Microservices{}, err
+	}
+
+	return spec, nil
+}
+
+func build(ctrlConfig configtempov1alpha1.ProjectConfig, params manifests.Params) ([]client.Object, error) {
+	// apply default values from Defaulter webhook
+	defaulterWebhook := v1alpha1.NewDefaulter(ctrlConfig.DefaultImages)
+	err := defaulterWebhook.Default(context.Background(), &params.Tempo)
+	if err != nil {
+		return nil, err
+	}
+
+	objects, err := manifests.BuildAll(params)
+	if err != nil {
+		return nil, err
+	}
+
+	return objects, nil
+}
+
+func toYAML(objects []client.Object, out io.Writer) error {
+	encoder := json.NewSerializerWithOptions(json.DefaultMetaFactory, nil, nil, json.SerializerOptions{Yaml: true})
+
+	for _, obj := range objects {
+		fmt.Fprintln(out, "---")
+		types, _, err := scheme.ObjectKinds(obj)
+		if err != nil {
+			log.Fatalf("error getting object kind: %v", err)
+		}
+		if len(types) == 0 {
+			log.Fatalf("could not find object kind for %v", obj)
+		}
+
+		obj.GetObjectKind().SetGroupVersionKind(types[0])
+		err = encoder.Encode(obj, out)
+		if err != nil {
+			log.Fatalf("error: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func usage() {
+	println("usage: go run ./cmd/generate/main.go -cr path/to/microservices/cr.yaml")
+}
+
+func main() {
```

Comment:
Alternatively we could also split up the main tempo-operator binary in multiple sub commands using a library like spf13/cobra.

Reply:
I would prefer that, it will be easier to use it (not additional build is required)

Reply:
I agree with this, having a sub commands will be better.

Reply:
@andreasgerstmayr are you going to do this change? WDYT?

Reply:
@rubenvp8510 yes, it's almost done. I just need to update a few places where the tempo-operator executable is started (in the manifests and Makefile).
I'll also add an e2e test and send a PR tomorrow.

Reply:
Done:
```
$ go run main.go                      
Usage:
  tempo-operator [command]

Available Commands:
  completion  Generate the autocompletion script for the specified shell
  generate    Generate YAML manifests from a Tempo CR
  help        Help about any command
  start       Start the Tempo operator

Flags:
      --config string   The controller will load its initial configuration from this file. Omit this flag to use the default configuration values. Command-line flags override configuration from this file.
  -h, --help            help for tempo-operator

Use "tempo-operator [command] --help" for more information about a command.
```

:)